shell/platform/linux/fl_standard_message_codec.cc

```diff
@@ -0,0 +1,525 @@
+
+static void write_uint8(GByteArray* buffer, uint8_t value) {
+  g_byte_array_append(buffer, &value, sizeof(uint8_t));
+}
+
+static void write_uint16(GByteArray* buffer, uint16_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(uint16_t));
+}
+
+static void write_uint32(GByteArray* buffer, uint32_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(uint32_t));
+}
+
+static void write_int32(GByteArray* buffer, int32_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(int32_t));
+}
+
+static void write_int64(GByteArray* buffer, int64_t value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(int64_t));
+}
+
+static void write_float64(GByteArray* buffer, double value) {
+  g_byte_array_append(buffer, reinterpret_cast<uint8_t*>(&value),
+                      sizeof(double));
+}
+
+// Checks there is enough data in the buffer to be read
+static gboolean check_size(GBytes* buffer,
+                           size_t* offset,
+                           size_t required,
+                           GError** error) {
+  if (*offset + required > g_bytes_get_size(buffer)) {
+    g_set_error(error, FL_MESSAGE_CODEC_ERROR,
+                FL_MESSAGE_CODEC_ERROR_OUT_OF_DATA, "Unexpected end of data");
+    return FALSE;
+  }
+  return TRUE;
+}
+
+// Gets a pointer to the current read location in the buffer
+static const uint8_t* get_data(GBytes* buffer, size_t* offset) {
+  return static_cast<const uint8_t*>(g_bytes_get_data(buffer, nullptr)) +
+         *offset;
+}
+
+// Functions to read FlValues in Standard encoding
+
+static gboolean read_uint8(GBytes* buffer,
+                           size_t* offset,
+                           uint8_t* value,
+                           GError** error) {
+  if (!check_size(buffer, offset, sizeof(uint8_t), error))
+    return FALSE;
+
+  *value = get_data(buffer, offset)[0];
+  (*offset)++;
+  return TRUE;
+}
+
+static gboolean read_uint16(GBytes* buffer,
+                            size_t* offset,
+                            uint16_t* value,
+                            GError** error) {
+  if (!check_size(buffer, offset, sizeof(uint16_t), error))
+    return FALSE;
+
+  *value = reinterpret_cast<const uint16_t*>(get_data(buffer, offset))[0];
+  *offset += sizeof(uint16_t);
+  return TRUE;
+}
+
+static gboolean read_uint32(GBytes* buffer,
+                            size_t* offset,
+                            uint32_t* value,
+                            GError** error) {
+  if (!check_size(buffer, offset, sizeof(uint32_t), error))
+    return FALSE;
+
+  *value = reinterpret_cast<const uint32_t*>(get_data(buffer, offset))[0];
+  *offset += sizeof(uint32_t);
+  return TRUE;
+}
+
+static FlValue* read_int32(GBytes* buffer, size_t* offset, GError** error) {
```

Comment:
Why do the first three `read_` methods use a totally different method for returning the value than the rest do? The inconsistency is very confusing.

Reply:
The rest are the high-level `FlValue` types, and the first three read C primitives (used for getting the object type and length fields). I could make a `FlReadBuffer` class that moved these out, but it didn't seem worth it at the time.

Reply:
Adding more comments has hopefully helped this a bit.

Reply:
How about adding `_as_value` (or maybe just `_value`) to all of these? I think it would be much clearer that `read_uint32` and `read_int32_as_value` are fundamentally different, while `read_uint32` and `read_int32` sound like they should be almost identical.